src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -1785,8 +1810,20 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
 
                     if (msbuildProject != null)
                     {
-                        msbuildProject.MSBuildNuGetProjectSystem.EndProcessing();
+                        if (isVsBatchStartEventRaised)
+                        {
+                            // raise batch end event only when we raised vs batch start
+                            msbuildProject.MSBuildNuGetProjectSystem.EndProcessing();
+                        }
+
+                        // raise nuget batch end event
+                        if (packageProjectEventArgs != null)
```

Comment:
it would be cleaner if the batch for the project system started and ended inside the this batch event instead of starting between the events and ending afterwards. For callers I think it makes it much easier to think about. This comment was on the last iteration also.